functional-samples/cookbook.geolocation-offscreen/service_worker.js

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+
+chrome.runtime.onMessage.addListener(handleMessages);
+
+let creating; // A global promise to avoid concurrency issues
+
+async function hasDocument() {
+  // Check all windows controlled by the service worker to see if one
+  // of them is the offscreen document with the given path
+  const matchedClients = await clients.matchAll();
+
+  return matchedClients.some((c) => c.url === OFFSCREEN_DOCUMENT_PATH);
+}
+
+async function setupOffscreenDocument(path) {
+  //if we do not have a document, we are already setup and can skip
+  if (!(await hasDocument())) {
+    // create offscreen document
+    if (creating) {
+      await creating;
+    } else {
+      creating = chrome.offscreen.createDocument({
+        url: path,
+        reasons: [
+          chrome.offscreen.Reason.GEOLOCATION ||
+            chrome.offscreen.Reason.DOM_SCRAPING
```

Comment:
What's the purpose / effect of using `||` here?

Is there some sort of case where `chrome.offscreen.Reason.GEOLOCATION` might not be defined? 🤔

Reply:
Hi @cuylerstuwe! Your hunch was correct. It is actually not in Chromium as of this moment. You can see the path [here](https://chromium-review.googlesource.com/c/chromium/src/+/4605111) though will be patched in ~tomorrow. Once added, it is the "correct" choice, but `DOM_SCRAPING` is the least incorrect choice for production code after that.

Reply:
It's interesting to see a Google-backed example of _"using the least-incorrect option"_.

I've wondered about this for a variety of use-cases, e.g., whether WebSockets would be allowed despite not having a reason specifically provided for them.

This seems to skew more in the favor of maybe using e.g. `DOM_SCRAPING` for WebSockets.

Reply:
Specifically in regard to reasons, I like to add cases as we find them. For the general case of "websockets", `DOM_SCRAPING` _may_ be the best option as of now. It really depends on what you are doing. I agree there needs to be a general websocket/real-time-communication reason added, however